Use a tree data structure to build template digests

tenderlove opened the pull request:
This PR changes the template digest code to build a dependency tree data structure before calculating the fingerprint and child finger prints.  I carefully made the tree produce exactly the same digests that the previous code produced (if you read the commits you'll see that I tested them side by side).
## Pros
- Slightly shorter implementation
- Easier to extend (we can just add new tree node types)
- Less memory consumption (I'll talk about this later)
- Easier to debug (I'll talk about that later too)
## Cons
- Possibly slower key calculation (I'll also talk about this later)
### Memory consumption / key calculation speed

The previous implementation would cache a digest for all nodes in any particular tree.  This would speed up digest calculation for templates that have common subtrees.  The downside is that any nodes that are not in common waste memory and once the app is warmed up, those keys are also wasting memory (since we only ever look up a digest by it's root node).  IOW, [our in-memory cache](https://github.com/rails/rails/blob/5971fddb7997231f3431c26ba17471ba52704852/actionview/lib/action_view/digestor.rb#L8) is always wasting memory.

The implementation I'm proposing here only caches the digest for the top most node (which is the node callers of `digest` were asking about).

Key calculation time should be amortized over the life of the app, so I don't think slowing down some calculations should be a big deal, but I'm happy to change if this turns out to be an issue.  Fortunately all we have to do to support that is pass the cache object down as we walk the tree.
### Strange things I found

If you have a wildcard dependency, like this:

```
<% # Template Dependency: users/* %>
```

_Everything_ in `users/` will be considered a "dependency", even files that are not templates.  Here is a visualization of the dependency tree for a test app:

![tree](http://i.imgur.com/jdZPg1v.png)

These files that are not templates will be considered as "missing" and the parent template will end up [getting an empty string returned for that template](https://github.com/rails/rails/blob/5971fddb7997231f3431c26ba17471ba52704852/actionview/lib/action_view/digestor.rb#L71) and [will digest a string like "--"](https://github.com/rails/rails/blob/5971fddb7997231f3431c26ba17471ba52704852/actionview/lib/action_view/digestor.rb#L114-L120).  The point being that these so-called "missing dependencies" can impact the digest of the parent template even though we don't care about them.  If I delete the file "users/show" from the tree in the picture, it would impact the digest for "users/new" even though "users/new" doesn't care whats inside "users/show" at all!  This test demonstrates the issue:

``` diff
+  def test_non_partials_do_not_impact_wildcard_dependencies
+    old_caching, ActionView::Resolver.caching = ActionView::Resolver.caching, false
+    before = digest "events/index"
+    add_template 'events/show'
+
+    ActionView::Digestor.cache.clear
+    assert_equal before, digest("events/index")
+  ensure
+    ActionView::Resolver.caching = old_caching
+  end
```

This test fails with the current implementation.  I would like to remove this "feature" and fix the above test.

If we're OK with this PR I'd also like to add a rake task that outputs the dependency tree as a dot file so we can get a visual representation of the deps.

tenderlove committed:
introduce a `tree` factory method for creating the dep tree

also add an EMPTY sentinel node

tenderlove committed:
pull template check up to match existing behavior

also remove the EMPTY node since we won't need it

tenderlove committed:
fix recursive templates

tenderlove committed:
clean up classes a little

tenderlove committed:
convert structs to a regular class

tenderlove committed:
change internal implementation to use a tree

tenderlove committed:
remove dead code

this commit removes unused code and changes the monitor to a mutex.
Since the digest doesn't recurse on itself anymore, we can just use a
mutex

tenderlove committed:
remove more dead code

tenderlove committed:
fix rake tasks

rafaelfranca commented on actionview/lib/action_view/digestor.rb:
```diff
@@ -25,108 +25,100 @@ class << self
       def digest(name:, finder:, **options)
         options.assert_valid_keys(:dependencies, :partial)
```
Using the kwargs for these since `options` can only have these two keys?

tenderlove replied:
Derp, ya, I'll change to kwargs.

rafaelfranca commented:
The implementation is easier and performance can be amortized as you said so for me it is :shipit:

tenderlove committed:
convert `digest` to only use kwargs

tenderlove commented:
I'm slightly concerned about changing the caching semantics before 5-0-stable is cut.  Our app doesn't use this yet (but I'm really excited to start) so I can't really get a good idea of the "real impact".

Another strange thing I noticed is that the original code will [disable cache on the template finder](https://github.com/rails/rails/blob/310e4418f4db899dcd07ff24f530ef8862c65f8b/actionview/lib/action_view/digestor.rb#L107).  I *think* the original intent was so that template changes will get picked up (it was originally introduced in 4725d588), but this code is not effective because once the top level node is cached, the templates dependencies are never traversed again.  If we're really concerned about cache key generation speed, we could stop disabling the template lookup cache, then we're only paying to md5 the template contents (which should be pretty darn fast).

/cc @jeremy

jeremy commented:
Could you mimic existing caching behavior, to start, then chop it after 5-0-stable?

tenderlove commented:
@jeremy ya, I could do that. Any idea when we're branching tho?

thedarkone commented:
Since we're no longer using `@@cache` to track recursion (using stack local `seen={}` instead), I think we can kill off `@@digest_mutex` altogether.

Something like this is now completely fine:
```ruby
# VERSION 1.
# let multiple threads potentially do redundant work
@@cache[cache_key] ||= tree(name, finder, dependencies).digest
# VERSION 2.
# or rely on TS::Map's concurrency primitive (no redundant work, allowing for possible concurrency on jruby/rbx for different cache_keys)
@@cache.compute_if_absent(cache_key) { tree(name, finder, dependencies).digest }
```
------

But, but... you promised to remove the recursion :crying_cat_face:!

tenderlove commented:
@thedarkone I mean recursion for *that* method! ;)  My love for functional programming won't allow me to remove all the recursion.

tenderlove committed:
Merge branch 'master' into treewip

* master: (27 commits)
  move digest cache on to the DetailsKey object
  remove object `hash` cache
  [ci skip] fix suggested change-replace 'an' with 'the in Rails engine guide'
  Missing documentation about hash algorithm option for MessageVerifier [ci skip]
  set `skip_listen` option to dummy appplication
  Fix the language in engines guide
  Add accidentally removed `#` [ci skip]
  fields_for_style needs to test for AC::Parameters
  Fix indentation for code block in changelog
  Remove accidentally duplicated change log title [ci skip]
  partially revert 69009f4473637a44ade26d954ef5ddea6ff903f2
  Remove needless `case_insensitive_comparison` in mysql2 adapter
  modify to `error` also abort when specify fail fast option
  Implement ActionController::Parameters#inspect
  remove unused method
  Remove unused Journey code
  Add Action Cable CHANGELOG in release notes [ci skip]
  Show proper error message when a non-relation object is passed to AR::Relation#or
  Fix semantics of test names for finish option in batches_test
  Fix typo
  ...

Conflicts:
	actionview/lib/action_view/digestor.rb

tenderlove committed:
cache all subnodes in the digest cache

this is for backwards compatibility.  We should remove this after
5-0-stable is created because digest calculation should be amortized.
Caching digests of subtrees will speed up digests for other trees that
share the same children.  However, this will also increase memory, and
after the app warms up, then those subtrees will never be reused which
means that memory is wasted.

tenderlove committed:
use the class level digest method for calculating the digest

tenderlove committed:
push injected dependencies up to the `digest` method

Only the root node has injected dependencies, so we don't need to care
about them at lower levels.  This change pushes the injected
dependencies up to where the user passed them in.

tenderlove committed:
push partial name regexp up

we know that all child templates will be considered as "partials", so
the only one that needs detection by name is the root node

tenderlove committed:
Merge branch 'master' into treewip

* master:
  eliminate warnings about multiple primary keys on habtm join tables
  Add methods to array delegation from `Relation`
  Fix button_to's params option to support nested names.
  Dependencies clean up
  Deprecate AC::Parameters#== with a Hash
  Fix AC::Parameters#== with other AC::Parameters
  Tests for AC::Parameters#==
  better docs for ActiveSupport::TestCase#assert_nothing_raised
  remove needless `debug_exception_response_format` config [ci skip]
  partial pass over the API guide [ci skip]
  Fix `unsubscribed` server side behavior
  Use a semaphore to signal message availability
  Fix master build
  Remove github gems from the master bug report templates
  Truncate ActionCable broadcast message to 300 chars
  Remove unused require

tenderlove committed:
Merge branch 'master' into treewip

* master: (113 commits)
  remove useless method
  Updated file documentation [ci skip]
  changes caching guide to add note on weak etags
  Don't put config.action_mailer.perform_caching entry twice in development.rb
  Fix wording and wrong reference
  Add Ruby formatting to CHANGELOG entry
  Fix ActionView's cache section reference
  Do not define methods in the included block
  Add caching guide in ActionMailer basics
  Add ActionMailer configuration options
  Preparing for 5.0.0.beta3 release
  Update 5.0 release notes
  Enable tmp_restart plugin for puma
  Prep release for Rails 5 beta3
  [ci skip] Move collection caching changelog entry.
  Ensure `drop_table` even if tests failure or interrupted
  :bomb: run the test @rafaelfranca :angry:
  Remove changelog entry for reverted commit
  Add CHANGELOG for https://github.com/rails/rails/pull/23734 [ci skip]
  No need CHANGELOG entry for #23849.
  ...
